pandas/core/groupby/base.py

```diff
@@ -3,13 +3,14 @@
 hold the whitelist of methods that are exposed on the
 SeriesGroupBy and the DataFrameGroupBy objects.
 """
-
-import types
-
-from pandas.util._decorators import make_signature
+from typing import TYPE_CHECKING, FrozenSet, Iterator, Type, Union
 
 from pandas.core.dtypes.common import is_list_like, is_scalar
 
+if TYPE_CHECKING:
```

Comment:
Was this actually causing circular imports? We import DataFrame and Series elsewhere within this directory didn't think those would cause issues.

Would like to avoid the TYPE_CHECKING variable generally

Reply:
Note this probably won't matter in light of other comments below

Reply:
No, it wasn't. Just my nervousness that it might cause one in a future commit.

Reply:
I don't understand you here. Could you be more explicit?

Reply:
where a type is expected, for example DataFrame, you can simply use the string 'DataFrame' rather than importing it (this only works for types :->) e.g.

```
def func(df: 'DataFrame) -> 'DataFrame':
   ...
```

is legit. Of course we prefer to import, but we need to avoid circular deps (e.g. in groupby in particular).

Reply:
Yes, it's legit, but it won't be infer correctly without the import?

At least in PyCharm it needs an import to recognize the type hint types do proper code completions, even when quoted.

Reply:
I've checked again, and actually if I remove the ``if TYPE_CHECKING`` statement, it will cause a circular import error.

So I think the if-statement must remain in order to do type checking. Also, the DataFrame rather than ABCDataFrame gives better results for PyCharm, without negativ results for mypy, so I propose keeping it that way.

Although adding type hints does feel like walking in uncharted territory as far as figuring out what best-practice is.

Reply:
Yea I think this will be required for "GroupBy" since it is not in the same module. FWIW you can do this by putting `reveal_type(base)` in the body of `whitelist_method_generator`.

Doing `Type['GroupBy']` yields the below errors:

```sh
$ mypy pandas/core/groupby/base.py
pandas/core/groupby/base.py:93: error: Name 'GroupBy' is not defined
pandas/core/groupby/base.py:124: error: Revealed type is 'Type[Any]'
```

Whereas this import for mypy infers correctly:

```sh
$ mypy pandas/core/groupby/base.py
pandas/core/groupby/base.py:127: error: Revealed type is 'Type[pandas.core.groupby.groupby.GroupBy]'
```

So I think this is correct for GroupBy (separate conversations still for DataFrame and Series)